File: Pairs/Pair_Sum_2.cpp
/*Given two unsorted arrays, find all pairs whose sum is a given number - O(max(n,m)) complexity*/

#include <bits/stdc++.h>
using namespace std;

class Answer{
    public:
        static void find_pairs(vector<int> v1, vector<int> v2, int sum){
            unordered_set<int> s;
            for(auto i = 0; i < v1.size(); i++)
                s.insert(v1[i]);
            for(auto j = 0; j < v2.size(); j++)
                if(s.find(sum-v2[j]) != s.end())
                    cout << sum-v2[j] << " " << v2[j] << endl;
        }
};

int main(){
    vector<int> tab1 = {1, 2, 3, 7, 5, 4 };
    vector<int> tab2 = {0, 7, 4, 3, 2, 1 };
    Answer::find_pairs(tab1,tab2,8);
}

File: Pairs/Pair_Sum2.py
# Given two unsorted arrays, find all pairs whose sum is equal to a given number
# O(max(N,M)) complexity

class Solution(object):
    def find_pairs(self,arr1:list, arr2:list, k:int):
        self.set = []
        self.set = set()
        for i in range(len(arr1)):
            self.set.add(arr1[i])
        for j in range(len(arr2)):
            if(k-arr2[j] in self.set):
                print(k-arr2[j],arr2[j])
                

if __name__=='__main__':
    arr1 = [1,2,3,7,5,4,8]
    arr2 = [0,7,4,3,2,1]
    k = 8
    s = Solution()
    s.find_pairs(arr1,arr2,k) # (8,0) (1,7) (4,4) (5,3) (7,1)
